great_expectations/data_context/store/store_backend.py

```diff
@@ -514,12 +454,181 @@ def get_random_hex(len=4):
+
+        if not os.path.isabs(root_directory):
+            raise ValueError("root_directory must be an absolute path. Got {0} instead.".format(root_directory))
+
+        self.root_directory = root_directory
+
+        self.full_base_directory = os.path.join(
+            self.root_directory,
+            self.base_directory,
+        )
+
+        safe_mmkdir(str(os.path.dirname(self.full_base_directory)))
+
+    def _get(self, key):
+        filepath = os.path.join(
+            self.full_base_directory,
+            self._convert_key_to_filepath(key)
+        )
+        with open(filepath) as infile:
+            return infile.read()
+
+    def _set(self, key, value):
+        filepath = os.path.join(
+            self.full_base_directory,
+            self._convert_key_to_filepath(key)
+        )
+        path, filename = os.path.split(filepath)
+
+        safe_mmkdir(str(path))
+        with open(filepath, "w") as outfile:
+            outfile.write(value)
+
+    def list_keys(self):
+        key_list = []
+        for root, dirs, files in os.walk(self.full_base_directory):
+            for file_ in files:
+                full_path, file_name = os.path.split(os.path.join(root, file_))
+                relative_path = os.path.relpath(
+                    full_path,
+                    self.full_base_directory,
+                )
+                if relative_path == ".":
+                    filepath = file_name
+                else:
+                    filepath = os.path.join(
+                        relative_path,
+                        file_name
+                    )
+
+                key_list.append(
+                    self._convert_filepath_to_key(filepath)
+                )
+
+        return key_list
+
+    def has_key(self, key):
+        assert isinstance(key, string_types)
+
+        all_keys = self.list_keys()
+        return key in all_keys
+
+
+class FixedLengthTupleS3StoreBackend(FixedLengthTupleStoreBackend):
+    """
+    Uses an S3 bucket as a store.
+
+    The key to this StoreBackend must be a tuple with fixed length equal to key_length.
+    The filepath_template is a string template used to convert the key to a filepath.
+    There's a bit of regex magic in _convert_filepath_to_key that reverses this process,
+    so that we can write AND read using filenames as keys.
+    """
+    def __init__(
+            self,
+            root_directory,
+            filepath_template,
+            key_length,
+            bucket,
+            prefix,
+            forbidden_substrings=["/", "\\"],
+            file_extension=None,
+            file_prefix=None,
+    ):
+        super().__init__(root_directory,
+                        filepath_template,
+                        key_length,
+                        forbidden_substrings=forbidden_substrings,
+                        file_extension=file_extension,
+                        file_prefix=file_prefix)
+        self.bucket = bucket
+        self.prefix = prefix
```

Comment:
IMO, we should deprecate `prefix`, `file_extension`, and `file_prefix`. All of these things can be specified more concisely and flexibly using `filepath_template`.

I think we should resolve this one way or another before merging.